code/components/jomjol_flowcontroll/MainFlowControl.cpp

```diff
@@ -474,6 +474,81 @@ esp_err_t handler_json(httpd_req_t *req)
     return ESP_OK;
 }
 
+/**
+ * Generates a http response containing the OpenMetrics (https://openmetrics.io/) text wire format 
+ * according to https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#text-format.
+ * 
+ * A MetricFamily with a Metric for each Sequence is provided. If no valid value is available, the metric is not provided.
+ * MetricPoints are provided without a timestamp. Additional metrics with some device information is also provided.
+ * 
+ * The default metric prefix is 'ai_on_the_edge_device_'.
+ * 
+ * example configuration for Prometheus (`prometheus.yml`):
+ * 
+ *    - job_name: watermeter
+ *      static_configs:
+ *        - targets: ['watermeter.fritz.box']
+ * 
+*/
+esp_err_t handler_openmetrics(httpd_req_t *req)
+{
+#ifdef DEBUG_DETAIL_ON
+    LogFile.WriteHeapInfo("handler_openmetrics - Start");
+#endif
+
+    ESP_LOGD(TAG, "handler_openmetrics uri: %s", req->uri);
+
+    if (bTaskAutoFlowCreated)
+    {
+        httpd_resp_set_hdr(req, "Access-Control-Allow-Origin", "*");
+        httpd_resp_set_type(req, "text/plain"); // application/openmetrics-text is not yet supported by prometheus
+
+        const string metricNamePrefix = "ai_on_the_edge_device_";   // FIXME: move to config-file
+
+        // get current measurement (flow)
+        string response = flowctrl.getOpenMetrics(metricNamePrefix);
+
+        // CPU Temperature
+        response += "# HELP " + metricNamePrefix + "cpu_temperature_celsius current cpu temperature in celsius\n# TYPE " 
```

Comment:
There are many string concatenations and repetitions. I contributed parts of [Tasmota's Prometheus endpoint](https://github.com/arendst/Tasmota/blob/v14.0.0/tasmota/tasmota_xsns_sensor/xsns_75_prometheus.ino) and found that moving such things to helper functions reduces the size of the generated code considerably (a concern on limited microcontrollers).

Reply:
I moved some of the business logic and helper functions into a separate component to improve readability. Code size was also reduced a bit.